### DISCUSSION TITLE
Work groups of different size

### DISCUSSION THREAD

**Discussion by mrzv** [AdaptiveCpp/AdaptiveCpp#649] [Q&A] [answered]
I have a question about SYCL, but this is the only place I know where I can ask it. (Plus it's about using this in hipSYCL.)

I'd like to use `parallel_for_work_group` to control what gets parallelized over compute units vs what gets vectorized within a compute unit. The trouble is that the underlying work that I want to do (which I think is supposed to be the size of the group) varies between groups. Is there any way to express this in SYCL? I.e., to invoke `group.parallel_for_work_item` with varying sizes?

As a toy example, imagine I have a sparse matrix, and I want to add up values in each column (or maybe double each value). I'd like to assign each column to a compute unit (using `parallel_for_work_group`), but then each column will have a different number of entries (from 0 to a billion). How can I launch the appropriate `parallel_for_work_item`? And does this even make sense? I.e., is this the right way to approach the problem in SYCL?

**Comment by illuhad** [comment, marked as answer]
You are very welcome to ask SYCL questions here, if you are fine with the fact that the answers you receive might be written through hipSYCL glasses ;) Other places to ask (and maybe receive more vendor-neutral answers) are the sycl tag on stackoverflow:
https://stackoverflow.com/questions/tagged/sycl
or the khronos devs slack:
https://khronosdevs.slack.com

First of all, I'd like to point out that using the hierarchical parallelism model (`parallel_for_work_group`/`parallel_for_work_item`) has been temporarily discouraged in the SYCL 2020 specification because the model was found to have issues around performance and implementability in some cases. It's likely that it will see major changes (aka breaking changes) in the next SYCL spec. The hipSYCL answer to those issues is our scoped parallelism v2 extension, which pioneers how a successor to the old hierarchical parallelism could look like:
https://github.com/illuhad/hipSYCL/pull/619

Additionally, hierarchical parallelism in hipSYCL is only partially implemented on GPU due to the massive implementation complexities and potential performance impact of implementing it "properly".

In hierarchical parallelism, you can submit `parallel_for_work_group` without specifying the work group size, and then passing the work group size to `parallel_for_work_item`. In this case, the iteration space of `parallel_for_work_item` can vary between multiple calls. This particular use case is called either flexible work group size, or logical work group size (depending on the revision of the SYCL spec you're looking at). However there is quite some nuance to this feature, and some functionality of hierarchical parallelism will not work anymore (or rather, work in unexpected ways) if you use it - in particular `private_memory`.

However, I would argue that you don't need this functionality, since you can easily build something similar without this nuance yourself. If your group size is something like a minimum vectorization width that you want to expose, you can just (pseudo code):
```c++
parallel_for_work_group(numgroups, vectorizationwidth, [=](auto group){
  int group_problem_size = ...
  for(int local_id_offset = 0; local_id_offset < group_problem_size; local_id_offset += vectorizationwidth) {
    // Will execute for vectorizationwidth items
    group.parallel_for_work_item([&](auto item){
       ...
    });
  }
});
```
This is effectively the same as what `parallel_for_work_item` with flexible group sizes would do.

Regarding your question about how to approach the problem in general, I would say that if you have greatly unbalanced work distribution across your work groups, this approach does not solve the fundamental problem of workload imbalance. Instead you would need to redistribute the actual parallel execution resources across your work such that the work is distributed evenly across, say, all SIMD lanes of all cores.
Typical approaches here could be performing a prepass to calculate the optimal distribution of work first in a separate kernel. This would also work well on GPUs. Of course, this is an additional overhead, so would only pay off if the imbalance is significant.

For less significant imbalance you could hope that the dynamic thread scheduling can balance out work group processing across multiple threads. However, if they are too small, you might run into runtime and scheduling overheads, and I'm not 100% sure how well this would work on GPU.

**Reply by mrzv** [reply]
Thank you for such a clear and thorough answer! Not only am I fine that the answer is through hipSYCL glasses, I prefer it that way.

I'll read up on the parallelism v2 extension. And many thanks for the code snippet; it's exactly what I was after!

Regarding load balancing, I have a follow-up question. It's not actually possible for me to estimate the amount of work required per column (per workgroup) — the sum or doubling example was just a toy. There is a great deal of variability in the amount of work required per column, but at least on the CPU, there is a clear advantage to parallelism, regardless of the imbalance. But there is a catch: it depends on how the outer for-loop is parallelized.

If the computation is simply broken into n/k equal chunks, where n is the total range of the outer loop (`numgroups`), and `k` is the number of "compute units", then there is no hope for good performance. On the other hand, if the work partition is finer grain than that — e.g., each group takes the next available work item — then the performance is much better. (This is not actually true, you really want each group to grab the next available "batch", but the point is that the batch is much smaller than n/k.)

My question is: how is `parallel_for` and `parallel_for_work_group` actually implemented? (On GPU. On CPU, I assume it's whatever OpenMP does, but actually I also don't know how that is implemented either.)

**Reply by illuhad** [reply]
Hi,

You are right that on CPU it mainly follows whatever OpenMP does, however there is some nuance depending on the exact type of the kernel invocation:
* `parallel_for` with `range` argument: `#pragma omp parallel` containing a single `#pragma omp for collapse(n)` across all work items with a single global iteration space.
* `parallel_for` with `nd_range` argument: This is much more complex because of the existence of collective barriers in this model, which prevents us from using a simple for loop across work items. Here, the idea is to use multi-threading across work groups and fibers for each work item. This boils down to a `#pragma omp parallel` with a static distribution of groups across threads. This is required AFAIR to avoid having to rebuild the fibers for each group, which is an important optimization.
* `parallel_for_work_group`: `pragma omp parallel` containing a single `#pragma omp for collapse(n)` to iterate across all work groups. `parallel_for_work_item` within groups is then implemented as `#pragma omp simd` loop.

On GPU, there is very little difference between these models, because common SYCL backends (OpenCL, CUDA, HIP etc) only have a single kernel model. Assuming a CUDA or HIP GPU, work groups are mapped to blocks, and work items to threads.

* In `parallel_for` with `nd_range` argument, hipSYCL just launches a HIP/CUDA kernel with the provided work group size. This is very straight-forward to implement.
* In `parallel_for` with `range` argument, the same thing happens except that the user does not provide information on the desired work group size, so hipSYCL selects a work group size (i.e. block size in HIP/CUDA) to execute the kernel with.
* GPUs generally don't support efficient mechanisms to spawn additional true parallelism dynamically, so in hipSYCL's incomplete `parallel_for_work_group` implementation, the idea is to launch it as a regular kernel. Consequently, all work items are there right from the start (this is similar to e.g. nested parallelism in Kokkos). `parallel_for_work_item` then effectively is a no-op that just executes the provided function object. If implemented properly, additional code would have to be inserted by the compiler to mask the execution of all work items that are not of local id 0 in the work group scope. This is currently lacking. When the user does not provide a work group size, hipSYCL selects one for the user. When a group size is passed to `parallel_for_work_item`, it will distribute that logical group range across the actual CUDA/HIP threads that are executing the work group.
